nav2_smoother/src/nav2_smoother.cpp

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2019 RoboTech Vision
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "nav2_core/exceptions.hpp"
+#include "nav2_smoother/nav2_smoother.hpp"
+#include "nav2_util/geometry_utils.hpp"
+#include "nav2_util/node_utils.hpp"
+#include "nav_2d_utils/conversions.hpp"
+#include "nav_2d_utils/tf_help.hpp"
+#include "tf2_ros/create_timer_ros.h"
+
+using namespace std::chrono_literals;
+
+namespace nav2_smoother {
+
+SmootherServer::SmootherServer()
+    : LifecycleNode("smoother_server", "", true),
+      lp_loader_("nav2_core", "nav2_core::Smoother"),
+      default_ids_{"SmoothPath"}, default_types_{
+                                      "nav2_smoother::CeresCostawareSmoother"} {
+  RCLCPP_INFO(get_logger(), "Creating smoother server");
+
+  declare_parameter("costmap_topic", rclcpp::ParameterValue(std::string(
+                                         "global_costmap/costmap_raw")));
+  declare_parameter("footprint_topic",
+                    rclcpp::ParameterValue(
+                        std::string("global_costmap/published_footprint")));
+  declare_parameter("global_frame", rclcpp::ParameterValue(std::string("map")));
+  declare_parameter("robot_base_frame",
+                    rclcpp::ParameterValue(std::string("base_link")));
+  declare_parameter("transform_tolerance", rclcpp::ParameterValue(0.1));
+  declare_parameter("smoother_plugins", default_ids_);
+
+  // // Launch a thread to run the costmap node
+  // costmap_thread_ = std::make_unique<nav2_util::NodeThread>(costmap_ros_);
+
+  RCLCPP_INFO(get_logger(), "Smoother server created");
+}
+
+SmootherServer::~SmootherServer() {
+  smoothers_.clear();
+  // costmap_thread_.reset();
+}
+
+nav2_util::CallbackReturn
+SmootherServer::on_configure(const rclcpp_lifecycle::State &) {
+  auto node = shared_from_this();
+
+  RCLCPP_INFO(get_logger(), "Configuring controller interface");
```

Comment:
Add log before getting the node from `shared_from_this`, that can statically crash if incorrectly used so it would be useful to know it entered this function beforehand in case a major refactor comes along and someone mishandles the node object.